Add Patch Manager dialog

has207 opened the pull request:
Simple interface that lets you enable/disable patches and persist the changes to disk.
Includes filter functionality to easily find specific games/patches if the user has many of them.

Implements #362

AdrianCassar commented:
It's a good start but needs improvement to become mergeable.

Suggestions:
Lock window movement and resizing
Centre window
Show different title for TUs or media ids
Don't log file paths
Fix conflicting ImGuI content ids
Fix the ImGuI assert on closure

has207 commented:
> It's a good start but needs improvement to become mergeable.

Thank you for the quick review! Comments inline

> Suggestions: Lock window movement and resizing Centre window 

done

> Show different title for TUs or media ids 

Could you clarify what you mean by this?

> Don't log file paths

Do you mean don't log file operations at all or specifically avoid showing file paths in the log?

> Fix conflicting ImGuI content ids

I made them more unique by prepending "patch", is that what you had in mind?

>Fix the ImGuI assert on closure

Added ImGui::End on close, I think that should address this?

has207 commented:
> > It's a good start but needs improvement to become mergeable.
> 
> Thank you for the quick review! Comments inline
> 
> > Suggestions: Lock window movement and resizing Centre window
> 
> done
> 
> > Show different title for TUs or media ids
> 
> Could you clarify what you mean by this?

Okay I figured it out, I was only testing on a couple patches I had locally, I downloaded all the ones from
game-patches and see how TU ones are handled. Should be showing that in the title now.

> 
> > Don't log file paths
> 
> Do you mean don't log file operations at all or specifically avoid showing file paths in the log?
> 
> > Fix conflicting ImGuI content ids
> 
> I made them more unique by prepending "patch", is that what you had in mind?
> 
> > Fix the ImGuI assert on closure
> 
> Added ImGui::End on close, I think that should address this?

AdrianCassar commented on src/xenia/app/patches_dialog.cc:
```diff
@@ -0,0 +1,367 @@
+ ******************************************************************************
+ * Xenia : Xbox 360 Emulator Research Project                                 *
+ ******************************************************************************
+ * Copyright 2024 Xenia Canary. All rights reserved.                          *
```
```suggestion
 * Copyright 2025 Xenia Canary. All rights reserved.                          *
```

AdrianCassar commented on src/xenia/app/patches_dialog.cc:
```diff
@@ -0,0 +1,367 @@
+
+  if (!std::filesystem::exists(patches_directory_)) {
+    std::filesystem::create_directories(patches_directory_);
+    XELOGI("Created patches directory: {}", patches_directory_.string());
```
If this includes absolute path then remove path.

has207 replied:
Just removed it entirely

AdrianCassar commented on src/xenia/app/patches_dialog.cc:
```diff
@@ -0,0 +1,367 @@
+  }
+
+  auto& all_patches = patch_db_->GetAllPatches();
+  XELOGI("Found {} patch files", all_patches.size());
```
```suggestion
```

AdrianCassar commented on src/xenia/app/patches_dialog.cc:
```diff
@@ -0,0 +1,367 @@
+    display.filename = patch_file.filename;
+
+    // Extract version info from filename (e.g., "(TU3)" or mark as base)
+    size_t tu_pos = display.filename.find("(TU");
```
Put this into a separate function

has207 replied:
done

AdrianCassar commented on src/xenia/app/patches_dialog.cc:
```diff
@@ -0,0 +1,367 @@
+}
+
+void PatchesDialog::SavePatchSettings() {
+  XELOGI("SavePatchSettings called, saving {} displays",
```
```suggestion
```

AdrianCassar commented on src/xenia/app/patches_dialog.cc:
```diff
@@ -0,0 +1,367 @@
+    std::filesystem::path patch_file_path =
+        patches_directory_ / display.filename;
+
+    XELOGI("Saving to patch file: {}", patch_file_path.string());
```
```suggestion
```

AdrianCassar commented on src/xenia/app/patches_dialog.cc:
```diff
@@ -0,0 +1,367 @@
+    XELOGI("Saving to patch file: {}", patch_file_path.string());
+
+    if (!std::filesystem::exists(patch_file_path)) {
+      XELOGW("Patch file not found: {}", patch_file_path.string());
```
```suggestion
```

AdrianCassar commented on src/xenia/app/patches_dialog.cc:
```diff
@@ -0,0 +1,367 @@
+            // Update or add the is_enabled field
+            auto* table = patch.as_table();
+            bool enabled = display.patches[index].is_enabled;
+            XELOGI("Setting patch {} '{}' is_enabled to {}", index,
```
Change format to `Show FPS enabled`

AdrianCassar commented on src/xenia/app/patches_dialog.cc:
```diff
@@ -0,0 +1,367 @@
+        file.close();
+        XELOGI("Saved patch settings for {:08X}", display.title_id);
+      } else {
+        XELOGE("Failed to open patch file for writing: {}",
```
```suggestion
```

AdrianCassar commented on src/xenia/app/patches_dialog.cc:
```diff
@@ -0,0 +1,367 @@
+      if (file.is_open()) {
+        file << config;
+        file.close();
+        XELOGI("Saved patch settings for {:08X}", display.title_id);
```
```suggestion
```

AdrianCassar commented on src/xenia/app/patches_dialog.cc:
```diff
@@ -0,0 +1,367 @@
+  ImGui::SetNextWindowPos(window_pos, ImGuiCond_Always);
+
+  bool window_open = true;
+  if (!ImGui::Begin("Patch Manager", &window_open,
```
add ImGuiWindowFlags_NoResize

has207 replied:
done

AdrianCassar commented on src/xenia/app/patches_dialog.cc:
```diff
@@ -0,0 +1,367 @@
+  ImGui::InputTextWithHint("##patch_filter", "Filter by title or patch name...",
+                           filter_text_.data(), filter_text_.size());
+
+  if (ImGui::Button("Refresh")) {
```
```suggestion
  if (ImGui::Button("Apply")) {
```

has207 replied:
Actually this button was superfluous, I cleaned up some silliness in how this was previously implemented and removed it

AdrianCassar commented on src/xenia/app/patches_dialog.cc:
```diff
@@ -0,0 +1,367 @@
+        }
+      }
+
+      ImGui::PushID(fmt::format("patch_game_{:08X}", display.title_id).c_str());
```
You want to use the hash here from within the patch file

has207 replied:
good call, done

AdrianCassar commented on src/xenia/app/patches_dialog.h:
```diff
@@ -0,0 +1,67 @@
+ ******************************************************************************
+ * Xenia : Xbox 360 Emulator Research Project                                 *
+ ******************************************************************************
+ * Copyright 2024 Xenia Canary. All rights reserved.                          *
```
```suggestion
 * Copyright 2025 Xenia Canary. All rights reserved.                          *
```

AdrianCassar commented on src/xenia/app/patches_dialog.h:
```diff
@@ -0,0 +1,67 @@
+  };
+
+  EmulatorWindow* emulator_window_;
+  std::vector<PatchDisplay> patch_displays_;
```
maybe a map would be better here?

Where uint32_t is title id and vector are patches for that title
```c++
std::map<uint32_t, std::vector<PatchInfo>>
```

has207 replied:
we never actually need to do anything with this datastructure other than iterate all the items so I think vector is more appropriate, also since we have the issue of dup title IDs due to TU variants the map actually becomes a bit awkward.

has207 commented:
I cleaned up the log spam and fixed up implementation to no longer do unnecessary reads/writes of the files, removing the pointless refresh functionality and added a bit of a nicer handling to the filter box giving it focus on create and adding a button to clear filter input.

PTAL!

has207 commented:
@AdrianCassar 

I simplified display title handling to just use the filename and not attempt to rely on title as defined in the patch file or extract TU version, etc. Turns out the format of these isn't very consistent and it would require some messy workarounds to support current set of patch files that will likely break for newly added ones in the future. So now I'm just listing all the patch files basically as they appear in the patches directory with only special processing for the title_id.

Similarly, this makes the grouping you suggested in discord (game->patch files > patch) also pretty awkward with existing patches as there are some that completely break the "game" convention, e.g. blue dragon and its jp variant which have completely different titles.

Anyway rather than try to work around the messy dataset I'm just surfacing it as is without too much bells and whistles outside of giving the user a way to easily search through it. Speaking of which, I got rid of the ugly X button on the search bar and just implemented Esc to clear which seems to be how ImGui wants you to do this sort of thing.

PTAL

Sowa95 commented:
Very much an edge case, but for some patches the value can be changed to achieve different results. This information is lost after passing thru manager.
I'm not a dev by the way. If it makes no sense to include such comments, then it is what it is.

58411403 as an example.
<img width="1854" height="462" alt="1" src="https://github.com/user-attachments/assets/07ca1766-7344-4e70-81b4-f7f2066fd5df" />

AdrianCassar commented:
> Very much an edge case, but for some patches the value can be changed to achieve different results. This information is lost after passing thru manager. I'm not a dev by the way. If it makes no sense to include such comments, then it is what it is.
> 
> 58411403 as an example. <img alt="1" width="1854" height="462" src="https://private-user-images.githubusercontent.com/113144725/485815805-07ca1766-7344-4e70-81b4-f7f2066fd5df.PNG?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTcwMTg3NzIsIm5iZiI6MTc1NzAxODQ3MiwicGF0aCI6Ii8xMTMxNDQ3MjUvNDg1ODE1ODA1LTA3Y2ExNzY2LTczNDQtNGU3MC04MWI0LWY3ZjIwNjZmZDVkZi5QTkc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwOTA0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDkwNFQyMDQxMTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mZDAzMjExYjRhNzQ1MWQxNTU4NDMwNDc2MjYxMTg4ZTNjNTRjNmY1MjNiYTRkZGRiMGIxNzhlZTYzZjkwODllJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.ao6n4J3CYI6feNRN1Jv9SmgKL6o_IUu7tHYTNt_jFQA">

Yeah it appears all comments are removed once a change has been written to the patch file, they should be preserved IMO.

has207 commented:
> > Very much an edge case, but for some patches the value can be changed to achieve different results. This information is lost after passing thru manager. I'm not a dev by the way. If it makes no sense to include such comments, then it is what it is.
> > 58411403 as an example. <img alt="1" width="1854" height="462" src="https://private-user-images.githubusercontent.com/113144725/485815805-07ca1766-7344-4e70-81b4-f7f2066fd5df.PNG?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTcwMTg3NzIsIm5iZiI6MTc1NzAxODQ3MiwicGF0aCI6Ii8xMTMxNDQ3MjUvNDg1ODE1ODA1LTA3Y2ExNzY2LTczNDQtNGU3MC04MWI0LWY3ZjIwNjZmZDVkZi5QTkc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwOTA0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDkwNFQyMDQxMTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mZDAzMjExYjRhNzQ1MWQxNTU4NDMwNDc2MjYxMTg4ZTNjNTRjNmY1MjNiYTRkZGRiMGIxNzhlZTYzZjkwODllJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.ao6n4J3CYI6feNRN1Jv9SmgKL6o_IUu7tHYTNt_jFQA">
> 
> Yeah it appears all comments are removed once a change has been written to the patch file, they should be preserved IMO.

That's a good catch, I'm not sure if the toml format supports some sort of one-of type option, but ideally this should be encoded into the patch properly so that it can be surfaced to the user through some sort of UI like this. Barring that, yes we should keep the comments. I'm traveling for the next few days and away from my computer so will take a look next week.

has207 commented:
> > > Very much an edge case, but for some patches the value can be changed to achieve different results. This information is lost after passing thru manager. I'm not a dev by the way. If it makes no sense to include such comments, then it is what it is.
> > > 58411403 as an example. <img alt="1" width="1854" height="462" src="https://private-user-images.githubusercontent.com/113144725/485815805-07ca1766-7344-4e70-81b4-f7f2066fd5df.PNG?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTcwMTg3NzIsIm5iZiI6MTc1NzAxODQ3MiwicGF0aCI6Ii8xMTMxNDQ3MjUvNDg1ODE1ODA1LTA3Y2ExNzY2LTczNDQtNGU3MC04MWI0LWY3ZjIwNjZmZDVkZi5QTkc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwOTA0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDkwNFQyMDQxMTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mZDAzMjExYjRhNzQ1MWQxNTU4NDMwNDc2MjYxMTg4ZTNjNTRjNmY1MjNiYTRkZGRiMGIxNzhlZTYzZjkwODllJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.ao6n4J3CYI6feNRN1Jv9SmgKL6o_IUu7tHYTNt_jFQA">
> > 
> > 
> > Yeah it appears all comments are removed once a change has been written to the patch file, they should be preserved IMO.
> 
> That's a good catch, I'm not sure if the toml format supports some sort of one-of type option, but ideally this should be encoded into the patch properly so that it can be surfaced to the user through some sort of UI like this. Barring that, yes we should keep the comments. I'm traveling for the next few days and away from my computer so will take a look next week.

Okay couldn't help myself before taking off, I implemented comment preservation.

PTAL

has207 committed:
Add Patch Manager dialog

Gliniak commented:
I'm not sure how much flexibility you consider is necessary, but I would rewrite whole file handling (read/write) to use tomlplusplus because these files (patches) are .toml files

has207 commented:
> I'm not sure how much flexibility you consider is necessary, but I would rewrite whole file handling (read/write) to use tomlplusplus because these files (patches) are .toml files

I squashed all the intermediate commits but that was actually the initial version. It's still using tomlplusplus for reading but for writing we have to preserve comments (see discussion above) so I updated the writes to only replace the bool values. There are alternate toml libs like toml11 that can do this more gracefully but would mean pulling in another dependency.

Gliniak commented:
Ohh ok. I thought it supports making comments, but seems like it doesn't. 
Ok I see no problem with it then from high level. I still need to check it later. 
Switching to another lib is not out of scope. I already did that once, so if switching to another won't require some major changes then I would prefer having newer solution.

has207 commented:
> Ohh ok. I thought it supports making comments, but seems like it doesn't. Ok I see no problem with it then from high level. I still need to check it later. Switching to another lib is not out of scope. I already did that once, so if switching to another won't require some major changes then I would prefer having newer solution.

Oh ok yeah I think that would be better as well, I was reluctant to pull in new deps willy nilly but if you're okay with that let me take a look

Gliniak commented:
I know it's asking a bit, but please make 3pp changes in separate PRs, also ideally all commits should have prefix [<something>] in case of 3pp [3PP], for kernel changes [Kernel] or [XAM] and so on. Thanks

has207 commented:
> I know it's asking a bit, but please make 3pp changes in separate PRs, also ideally all commits should have prefix [] in case of 3pp [3PP], for kernel changes [Kernel] or [XAM] and so on. Thanks

It looks like toml11 doesn't compile on windows right now, it's probably not a big fix on their end, but I'm not sure I want to keep running in circles on this. I'll close this PR
